source/host_resolver.c

```diff
@@ -732,38 +732,64 @@ static int default_resolve_host(
 
     struct aws_host_address *aaaa_record = aws_lru_cache_use_lru_element(&host_entry->aaaa_records);
     struct aws_host_address *a_record = aws_lru_cache_use_lru_element(&host_entry->a_records);
+    struct aws_host_address address_array[2];
+    AWS_ZERO_ARRAY(address_array);
+    struct aws_array_list callback_address_list;
+    aws_array_list_init_static(&callback_address_list, address_array, 2, sizeof(struct aws_host_address));
 
     if ((aaaa_record || a_record) && host_entry->keep_active) {
         AWS_LOGF_DEBUG(
             AWS_LS_IO_DNS,
             "id=%p: cached entries found for %s returning to caller.",
             (void *)resolver,
             host_name->bytes);
-        struct aws_host_address *address_array[2];
-        AWS_ZERO_ARRAY(address_array);
-        struct aws_array_list callback_address_list;
-        aws_array_list_init_static(&callback_address_list, address_array, 2, sizeof(struct aws_host_address *));
 
+        /* these will all need to be copied so that we don't hold the lock during the callback. */
         if (aaaa_record) {
-            aws_array_list_push_back(&callback_address_list, &aaaa_record);
-            AWS_LOGF_TRACE(
-                AWS_LS_IO_DNS,
-                "id=%p: vending address %s for host %s to caller",
-                (void *)resolver,
-                aaaa_record->address->bytes,
-                host_entry->host_name->bytes);
+            struct aws_host_address aaaa_record_cpy;
+            if (!aws_host_address_copy(aaaa_record, &aaaa_record_cpy)) {
+                aws_array_list_push_back(&callback_address_list, &aaaa_record_cpy);
+                AWS_LOGF_TRACE(
+                    AWS_LS_IO_DNS,
+                    "id=%p: vending address %s for host %s to caller",
+                    (void *)resolver,
+                    aaaa_record->address->bytes,
+                    host_entry->host_name->bytes);
+            }
         }
         if (a_record) {
-            aws_array_list_push_back(&callback_address_list, &a_record);
-            AWS_LOGF_TRACE(
-                AWS_LS_IO_DNS,
-                "id=%p: vending address %s for host %s to caller",
-                (void *)resolver,
-                a_record->address->bytes,
-                host_entry->host_name->bytes);
+            struct aws_host_address a_record_cpy;
+            if (!aws_host_address_copy(a_record, &a_record_cpy)) {
+                aws_array_list_push_back(&callback_address_list, &a_record_cpy);
+                AWS_LOGF_TRACE(
```

Comment:
Is it worthwhile to move this outside the lock and log the copies so we aren't logging inside a lock?

Reply:
probably, though there's an allocation in that copy. And this isn't really a hot path (this almost always comes from the caller's thread.